File: src/main/java/org/vanautrui/languages/parsing/astnodes/terminal/IdentifierNode.java
package org.vanautrui.languages.parsing.astnodes.terminal;

import org.vanautrui.languages.lexing.collections.TokenList;
import org.vanautrui.languages.lexing.tokens.utils.Token;
import org.vanautrui.languages.lexing.tokens.IdentifierToken;
import org.vanautrui.languages.parsing.IASTNode;

public class IdentifierNode implements IASTNode {

    //this identifies a variable

    public String name;

    public IdentifierNode(TokenList tokens) throws Exception {

        //List<DragonToken> copy = new ArrayList<>(tokens);

        //Pattern p = Pattern.compile(IdentifierToken.regex_alphanumeric_identifier);

        Token token = tokens.get(0);

        if (token instanceof IdentifierToken) {
            this.name = ((IdentifierToken) token).getContents();
            tokens.consume(1);

        } else {
            throw new Exception("could not read identifier");
        }

    }

    @Override
    public String toSourceCode() {
        return this.name;
    }

}

File: dragon-parser/src/main/compiler/parsing/astnodes/terminal/IdentifierNode.cpp

//standard headers
#include <vector>
#include <set>

//project headers
#include "IdentifierNode.hpp"

class IdentifierNode : IASTNode {

public:
	//this identifies a variable
	string identifier;

	IdentifierNode(TokenList tokens) throws Exception {

		IToken token = tokens.get(0);

		if (token instanceof IdentifierToken) {
			this.identifier = token.getContents();
			tokens.consume(1);

		} else {
			String msg = "could not read identifier from token: " + token.getContents();
			msg += " in " + tokens.relPath + ":" + token.getLineNumber();
			msg += " in context '" + tokens.toSourceCodeFragment().substring(0, Math.min(20, tokens.toSourceCodeFragment().length())) + "'";

			throw new Exception(msg);
		}

	}

};